src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala

```diff
@@ -142,6 +149,94 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
     }
   }
 
+  // collect variables used in the expression
+  def collectVariables(expr: Expr) : Set[String] = {
+    var set = new HashSet[String]()
+    def collectVariablesInner(e: Expr) : Unit = e match {
+      case VarExpr(name) => set += name
+      case FuncExpr(function, args, agg) => args foreach (collectVariablesInner(_))
+      case BinaryOpExpr(lhs, op, rhs) => {
+        collectVariablesInner(lhs)
+        collectVariablesInner(rhs)
+      }
+      case TypecastExpr(lhs, rhs) => collectVariablesInner(lhs)
+      case _ =>
+    }
+    collectVariablesInner(expr)
+    set
+  }
+
+  def collectVariables(pattern: Pattern, definitionOnly: Boolean) : Set[String] = {
+    var set = new HashSet[String]()
+    pattern match {
+      case VarPattern(name) => set += name
+      case ExprPattern(e) => if (!definitionOnly) set ++= collectVariables(e)
+      case PlaceholderPattern() =>
+    }
+    set
+  }
+
+  def collectVariables(cond: Cond) : Set[String] = {
+    var set = new HashSet[String]()
+    cond match {
+      case ComparisonCond(lhs, op, rhs) => {
+        set ++= collectVariables(lhs)
+        set ++= collectVariables(rhs)
+      }
+      case CompoundCond(lhs, op, rhs) => {
+        set ++= collectVariables(lhs)
+        set ++= collectVariables(rhs)
+      }
+      case NegationCond(c) => set ++= collectVariables(c)
+    }
+    set
+  }
+
+  // collect variable
+  def collectVariablesFromBody(body: Body, definitionOnly: Boolean) : Set[String] = {
+    var set = new HashSet[String]()
+    body match {
+      case a: BodyAtom => a.terms.foreach { x =>
+        set ++= collectVariables(x, definitionOnly)
+      }
+      case a: QuantifiedBody => a.bodies.foreach {
+        x => set ++= collectVariablesFromBody(x, definitionOnly)
+      }
+      case a: Cond => if (!definitionOnly) set ++= collectVariables(a)
+    }
+    set
+  }
+
+  // check if variables have bindings in the body
+  def checkVariableBindings(stmt: Statement) {
+    // check variable bindings in a conjunctive query
+    def checkCq(cq: ConjunctiveQuery, additionalVars: Set[String] = Set()) {
+      // collect variable definitions and usages
+      var varDefSet = new HashSet[String]()
+      var varUseSet = new HashSet[String]()
+      cq.bodies.foreach { case x =>
+        x.foreach { case a =>
+          varDefSet ++= collectVariablesFromBody(a, true)
+          varUseSet ++= collectVariablesFromBody(a, false)
+        }
+      }
+      cq.head.terms.foreach  { x => varUseSet ++= collectVariables(x) }
+      varUseSet ++= additionalVars
+      varUseSet.foreach { x =>
+        if (!varDefSet.contains(x)) error(stmt, s"Variable ${x} does not have bindings")
```

Comment:
I think it's better to check if `varUseSet - varDefSet` is empty and show all undefined variables. (and perhaps also warn unused variables with a way to print warning messages to stderr?)